classes/decision/types/traits/withReviewRound.inc.php

```diff
@@ -0,0 +1,234 @@
+<?php
+
+/**
+ * @file classes/decision/types/traits/withReviewRound.inc.php
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
+ * Copyright (c) 2000-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class decision
+ *
+ * @brief Helper functions to provide review round related details associated with submission
+ */
+
+namespace PKP\decision\types\traits;
+
+use APP\submission\Submission;
+use Exception;
+use PKP\db\DAORegistry;
+use PKP\submission\reviewAssignment\ReviewAssignment;
+use PKP\submission\reviewAssignment\ReviewAssignmentDAO;
+
+trait withReviewRound
```

Comment:
I also thought about using it together with `InExternalReviewRound` and `InInternalReviewRound` and tried to use it as a trait used by those 2 other traits . Problem is that we need this trait `InReviewRound` even in other decision class which is not in any kind of review stage , for example the case such as `Back from the Copyediting Stage` . in that case , it may not be associated with review stage at all but we do need to check it so that we do not pass it to submission directly if there is any review stage between those 2 . that is why I could not combine them together . basically the `InReviewRound` trait provide few methods to do some checking if there is any association to any review stage/round .

Reply:
I guess what I'm struggling with is that `withReviewRound` doesn't seem to have a clear scope. Most of its methods are used in one decision or another. But they aren't all used in each decision, for example. I wonder if it should be broken up into a couple of different traits? 

Here are some of the things I'm wondering about:

1. The methods `isOnlyReviewRound()`, `hasMultipleReviewRound()`, and `hasReviewRound()` all feel like they could be one method, `getReviewRoundCount()`. So you'd use `getReviewRoundCount() === 1`, `getReviewRoundCount() > 1` and `getReviewRoundCount() >= 1`.
2. The methods `getCompletedReviewerIds()` and `getActiveReviewersId()` are only used in `NotifyReviewers`, but there is no clear dependency on `withReviewRound()`. Why not implement these directly in `NotifyReviewers`?
3. You have a base method, `getRereviewAssignments()`, that's used by several other methods. However, this base method doesn't save much code. As far as I can see, it only really saves a single call to a DAO. Since the code is already isolated in the DAO, do we need to encapsulate it in another method here? I wonder if, without this extra layer, it would make sense to keep `getCompletedReviewAssignments()` and `getActiveReviewAssignments()` in the `NotifyReviewers` trait?
4. It doesn't seem like `getConfirmedReviewerAssignments()` is used?

Reply:
Done .